lib/matplotlib/layoutbox.py

```diff
@@ -0,0 +1,727 @@
+        # bottom
+        if not sol.hasEditVariable(self.bottom_margin_min):
+            sol.addEditVariable(self.bottom_margin_min, 'strong')
+            sol.suggestValue(self.bottom_margin_min, 0.0001)
+        # if not sol.hasEditVariable(self.bottom_margin):
+        #     sol.addEditVariable(self.bottom_margin, 'weak')
+        #     sol.suggestValue(self.bottom_margin, 0.0001)
+        c = (self.bottom_margin == self.bottom - self.parent.bottom)
+        self.solver.addConstraint(c | 'required')
+        c = (self.bottom_margin >= self.bottom_margin_min)
+        self.solver.addConstraint(c | 'required')
+        # top
+        if not sol.hasEditVariable(self.top_margin_min):
+            sol.addEditVariable(self.top_margin_min, 'strong')
+            sol.suggestValue(self.top_margin_min, 0.0001)
+        # if not sol.hasEditVariable(self.top_margin):
+        #     sol.addEditVariable(self.top_margin, 'weak')
+        #     sol.suggestValue(self.top_margin, 0.0001)
+        c = (self.top_margin == self.parent.top - self.top)
+        self.solver.addConstraint(c | 'required')
+        c = (self.top_margin >= self.top_margin_min)
+        self.solver.addConstraint(c | 'required')
+
+    def add_child(self, child):
+        self.children += [child]
+
+    def remove_child(self, child):
+        try:
+            self.children.remove(child)
+        except ValueError:
+            print("Tried to remove child that doesn't belong to parent")
+
+    def add_constraints(self):
+        sol = self.solver
+        # never let width and height go negative.
+        for i in [self.min_width, self.min_height]:
+            sol.addEditVariable(i, 1e9)
+            sol.suggestValue(i, 0.0)
+        # define relation ships between things thing width and right and left
+        self.hard_constraints()
+        # self.soft_constraints()
+        if self.parent:
+            self.parent_constrain()
+        # sol.updateVariables()
+
+    def parent_constrain(self):
+        parent = self.parent
+        eps = 0.0000000
+        hc = [self.left >= parent.left + eps,
+              self.bottom >= parent.bottom + eps,
+              self.top <= parent.top - eps,
+              self.right <= parent.right - eps]
+        for c in hc:
+            self.solver.addConstraint(c | 'required')
+
+    def hard_constraints(self):
+        hc = [self.width == self.right - self.left,
+              self.height == self.top - self.bottom,
+              self.h_center == (self.left + self.right) * 0.5,
+              self.v_center == (self.top + self.bottom) * 0.5,
+              self.width >= self.min_width,
+              self.height >= self.min_height]
+        for c in hc:
+            self.solver.addConstraint(c | 'required')
+
+    def soft_constraints(self):
+        sol = self.solver
+        if self.tightwidth:
+            suggest = 0.
+        else:
+            suggest = 20.
+        c = [(self.pref_width == suggest)]
```

Comment:
no need for parentheses here and below (you don't have them in constrain_geometry either)